### PR TITLE
TimeLock Agents to Develop

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Please use "Rebase and Merge" for this PR when merging it into develop. Thanks.**

- [x] merge conflict
- [x] release notes
- [x] concretize the agent

**Goals (and why)**:
Merge the timelock agent branch to develop.
This refactors the creation logic of TimeLock significantly, to avoid depending on anything Dropwizard-related in particular.

**Implementation Description (bullets)**:
- Push all Dropwizard specific logic into `timelock-server`.

**Concerns (what feedback would you like?)**:
- This is merging a feature branch down to develop. I don't know if there are any landmines here.
- The `PaxosTimeLockServer` still exists in the codebase but is not being used. I'm going ahead with this first as #2174 is somewhat lower priority.
- Do we want to release 0.53 first, if we're keen on pushing out #2172 soon?

**Where should we start reviewing?**:
Probably for this one a quick scan will do - this is #2141, #2156, #2168, #2237 plus merge conflicts and release notes.

Actually `release-notes.rst` is probably a good one.

**Priority (whenever / two weeks / yesterday)**: soon, BUT maybe after 0.53 release.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2247)

### CODE REVIEW

**Comment by nziebart** [comment]
At this point is PaxosTimelockServer just sitting around unused?

**Comment by jeremyk-91** [comment]
@nziebart yes, that's right.

**Comment by tpetracca** [comment]
Piling on to @nziebart's comments with the weirdness I experienced attempting to use this internally just now. I think now that we force you to use paxos we should:
- rename `algorithm` in the configs to `paxos`
- delete the `PaxosInstallConfiguration.createTimeLockAgent` method
- converge `TimeLockAgent` and `PaxosAgent` into one concrete class called `TimeLockAgent`

This way the consumer of this stuff simply does
```
TimeLockAgent agent = new TimeLockAgent(install, runtime, registrar);
agent.createAndRegisterResources();
```
instead of
```
TimeLockAgent agent = install.timelock().algorithm().createTimeLockAgent(
        install,
        runtime,
        ImmutableTimeLockDeprecatedConfiguration.builder().build(),
        registrar);
agent.createAndRegisterResources();
```

File: timelock-agent/src/main/java/com/palantir/timelock/paxos/PaxosAgent.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        // Finally, register the endpoints associated with the clients.
+        super.createAndRegisterResources();
+    }
+
+    // No runtime configuration at the moment.
+    private void registerPaxosResource() {
+        registrar.accept(paxosResource);
+    }
+
+    private void registerPaxosExceptionMappers() {
```
Comment (jeremyk-91) [inline timelock-agent/src/main/java/com/palantir/timelock/paxos/PaxosAgent.java]: Yep, will move this. Following @tpetracca's later comment I think it makes sense to converge the generic timelock agent and concrete paxos agent.

File: timelock-agent/src/main/java/com/palantir/timelock/paxos/PaxosAgent.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    @Override
+    public void createAndRegisterResources() {
+        registerPaxosResource();
+        registerPaxosExceptionMappers();
+        leadershipCreator.registerLeaderElectionService();
+
+        // Finally, register the endpoints associated with the clients.
+        super.createAndRegisterResources();
```
Comment (jeremyk-91) [inline timelock-agent/src/main/java/com/palantir/timelock/paxos/PaxosAgent.java]: Yep, I wasn't too happy with this part of the code either. I'm thinking of condensing everything into one concrete class actually, given that we don't allow people to pick anything other than paxos.

**Comment by jeremyk-91** [comment]
This one's pending separate commit to converge the concrete paxos agent and abstract timelock agent down into 1 concrete class.

**Comment by jeremyk-91** [comment]
@nziebart @tpetracca This one should be good to go again.

**Comment by nziebart** [comment]
going to start working with this in @tpetracca'a internal project - will follow up if i run into any issues

**Comment by jeremyk-91** [comment]
Merging. Thanks for testing @nziebart!

**Comment by jeremyk-91** [comment]
Bleh. I'll do a normal merge, since Rebase-And-Merge is broken due to conflicts and we do want to keep the history

**Pull request merged**